tools/PI/DevHome.PI/Views/BarWindowHorizontal.xaml.cs

```diff
@@ -179,60 +183,79 @@ public void PopulateCommandBar()
         }
     }
 
-    private void AddToolToCommandBar(ExternalTool tool)
+    private AppBarButton CreateAppBarButton(ExternalTool tool, PinOption pinOption)
     {
         AppBarButton button = new AppBarButton
         {
             Label = tool.Name,
             Tag = tool,
         };
 
-        button.Icon = tool.MenuIcon;
+        button.Icon = new ImageIcon
+        {
+            Source = tool.ToolIcon,
+        };
 
         button.Click += _viewModel.ExternalToolButton_Click;
+        button.ContextFlyout = CreateMenuFlyout(tool, pinOption);
+
+        ToolTipService.SetToolTip(button, tool.Name);
 
+        return button;
+    }
+
+    private MenuFlyout CreateMenuFlyout(ExternalTool tool, PinOption pinOption)
+    {
         MenuFlyout menu = new MenuFlyout();
-        menu.Items.Add(tool.IsPinned ? CreateUnPinMenuItem(tool) : CreatePinMenuItem(tool));
+        menu.Items.Add(pinOption == PinOption.Pin ? CreatePinMenuItem(tool) : CreateUnPinMenuItem(tool));
         menu.Items.Add(CreateUnregisterMenuItem(tool));
-        button.ContextFlyout = menu;
 
-        // If a tool is pinned, we'll add it to the primary commands list, otherwise the secondary commands list
+        return menu;
+    }
+
+    private void AddToolToCommandBar(ExternalTool tool)
+    {
+        // We create 2 copies of the button, one for the primary commands list and one for the secondary commands list.
+        // We're not allowed to put the same button in both lists.
+        AppBarButton pinnedButton = CreateAppBarButton(tool, PinOption.UnPin); // The pinned button should always have the unpin option

```

Comment:
>pinnedButton [](http://example.com/codeflow?start=21&length=12)

maybe rename to primaryComamndListButton and secondaryCommandListButton?

Reply:
Sounds good.